docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,227 @@
+descriptor set 0 and binding 2, the DescriptorTableVsPtr must be loaded first.
+
+Suppose the compiler is generating code to access descriptor set 1 at binding 0
+in a standard compile, where the user data nodes are available. Because it is a
+top level node, the compiler will try to find an input register to hold the
+descriptor. Say it finds `s3`, and it is compiling the vertex shader. Then in
+the PAL metadata the entry `SPI_SHADER_USER_DATA_VS_3` will be assigned the
+offset of the descriptor, which is `0` in this case. Then that descriptor will
+be available in `s3`. Note that the only resource descriptor that can be in the
+top level are buffer descriptors.
+
+Now suppose that the compiler could not find an input registers. Then the
+resource descriptor will be spilled. That is, it will be stored in memory. One
+of the input registers will be the address of the spill table. The descriptor is
+loaded at the given offset in the spill table.
+
+Now suppose the compiler must generate code to access descriptor 0 at binding 1
+in a standard compile, where the user data nodes are available. The compiler
+will first have to generate code to load the DescriptorTableVsPtr. This is
+similar to loading descriptor set 1 at binding 0. Then the offset of the
+descriptor is added to that.
+
+An unlinked shader compilation has to account for both a top level descriptor or
+a descriptor in a descriptor table. An example can be found in the
+[buffer descriptor load test](../llpc/test/shaderdb/relocatable_shaders/DescBufferReloc.vert).
+
+Note that for the top level case, it is assumed that the descriptor has been
+spilt. The code is generated this way in
+[`DescBuilder::CreateLoadBufferDesc`](../lgc/builder/DescBuilder.cpp). The case
+where `node == topeNode` cannot be reached for unlinked shaders with missing
+user data nodes. Then when replacing the `reloc::DescriptorUseSpillTable`
+relocation in [`RelocHandler::getValue`](../lgc/elfLinker/RelocHandler.cpp), the
+descriptor is spilled if it is a top level node.
+
+Also note that it reserves two input registers, one of which is linked to
+descriptor set 0. That register entry does not contain an offset into user data
+nodes as PAL expected. During the link step, the value will be replaced with the
+offset of the descriptor table that contains descriptor set 0. This happens in
+[`PalMetadata::fixUpRegisters`](../lgc/state/PalMetadata.cpp).
+
+## Compact buffer descriptors
+
+While building an unlinked shader, it is not possible to know if the compact
+buffer descriptors will be used or not. It is unclear if it would be worth
+generating code general enough to work in both cases. The extra overhead would
+potentially negate any gains from using compact buffer descriptors. Therefore
+unlinked shaders are compiled assuming that there are no compact buffer
+descriptors. If the XGL option `EnableRelocatableShaders` is true, then the
+driver will not use compact buffer descriptors.
```

Comment:
Perhaps add a background note that a compact descriptor is only ever in the top-level table, and the driver puts descriptors in the top-level table for a "dynamic descriptor", to implement the Vulkan functionality where the app can modify the start offset of a buffer without rebuilding the whole pipeline state.

Reply:
I added some of this to the previous section where I define "top-level" nodes.  I continue to use top-level instead of dynamic descriptors because that is what is seen in the code.